src/Features/CSharp/Portable/ConvertProgram/ConvertProgramHelpers.cs

```diff
@@ -0,0 +1,121 @@
+                if (programType != null)
+                {
+                    if (programType.GetMembers(WellKnownMemberNames.TopLevelStatementsEntryPointMethodName).FirstOrDefault() is IMethodSymbol mainMethod)
+                    {
+                        var classDeclaration = await GenerateProgramClassAsync(
+                            document, programType, mainMethod, cancellationToken).ConfigureAwait(false);
+
+                        var newRoot = root.RemoveNodes(members.OfType<GlobalStatementSyntax>().Skip(1), SyntaxGenerator.DefaultRemoveOptions);
+                        Contract.ThrowIfNull(newRoot);
+
+                        var firstGlobalStatement = newRoot.Members.OfType<GlobalStatementSyntax>().Single();
+                        newRoot = newRoot.ReplaceNode(
+                            firstGlobalStatement,
+                            FixupComments(classDeclaration.WithLeadingTrivia(firstGlobalStatement.GetLeadingTrivia())));
+
+                        return document.WithSyntaxRoot(newRoot);
+                    }
+                }
+            }
+
+            return document;
+        }
+
+        private static async Task<ClassDeclarationSyntax> GenerateProgramClassAsync(
+            Document document,
+            INamedTypeSymbol programType,
+            IMethodSymbol mainMethod,
+            CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var option = options.GetOption(CodeStyleOptions2.RequireAccessibilityModifiers);
+            var accessibilityModifiersRequired = option.Value is AccessibilityModifiersRequired.ForNonInterfaceMembers or AccessibilityModifiersRequired.Always;
+
+            var generator = document.GetLanguageService<SyntaxGenerator>();
+
+            // See if we have an existing part in another file.  If so, we'll have to generate our declaration as partial.
+            var hasExistingPart = programType.DeclaringSyntaxReferences.Any(d => d.GetSyntax(cancellationToken) is TypeDeclarationSyntax);
+
+            var root = (CompilationUnitSyntax)await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            return (ClassDeclarationSyntax)generator.ClassDeclaration(
+                WellKnownMemberNames.TopLevelStatementsEntryPointTypeName,
+                accessibility: accessibilityModifiersRequired ? Accessibility.Internal : Accessibility.NotApplicable,
+                modifiers: hasExistingPart ? DeclarationModifiers.Partial : DeclarationModifiers.None,
+                members: new[]
+                {
+                    (MemberDeclarationSyntax)generator.WithAccessibility(
+                        generator.MethodDeclaration(mainMethod, "Main", GetStatements(root)),
+                        accessibilityModifiersRequired ? Accessibility.Private : Accessibility.NotApplicable)
```

Comment:
It may have been async

Reply:
async is handled already by looking at what the mainMethod was.  see the tests :)